cstore_writer.c

```diff
@@ -318,96 +326,221 @@ CStoreWriteRow(TableWriteState *writeState, Datum *columnValues, bool *columnNul
+
+		/*
+		 * We here we decide where to start writing new table footer metadata.
+		 * If deserialization fail for any reason starting block is set to an invalid
+		 * starting block number (0).
+		 *
+		 * if there is a parsing error we start from the first buffer (1). If it is fine
+		 * we check if there are enough empty buffers to accommodate all blocks of
+		 * table footer metadata and set new starting buffer to first buffer (1).
+		 * If there are not enough buffers we start writing after the all used buffers.
+		 */
+		if (startingBlock == 0)
+		{
+			newStartingBlock = 1;
+		}
+		else if (actualBlockCount < startingBlock)
+		{
+			newStartingBlock = 1;
+		}
+		else
+		{
+			newStartingBlock = startingBlock + blockCount;
+		}
+	}
+
+	dataOffset = 0;
+	memcpy(wholeFooter->data, (const char*) &dataLength, sizeof(int32));
 
-	SyncAndCloseFile(tableFooterFile);
+	for (currentBlockNumber = 0; currentBlockNumber < actualBlockCount; currentBlockNumber++)
+	{
+		Buffer buffer = InvalidBuffer;
+		BlockNumber blockNumber = currentBlockNumber + newStartingBlock;
+		Page page;
+		PageHeader pageHeader = NULL;
+		char * pageData = NULL;
+		BlockNumber actualBlockNumber = InvalidBlockNumber;
+
+
+		if (blockNumber >= originalBlockCount)
+		{
+			blockNumber = P_NEW;
+		}
+
+		buffer = ReadBufferExtended(relation, FOOTER_FORKNUM, blockNumber, RBM_NORMAL, NULL);
+
+		LockBuffer(buffer, BUFFER_LOCK_EXCLUSIVE);
+
+		START_CRIT_SECTION();
+		page = BufferGetPage(buffer);
+
+		PageInit(page, BLCKSZ, 0);
+		pageData = PageGetContents(page);
+
+		int copySize = blockDataSize;
+
+		if ( (dataLength - dataOffset) <= blockDataSize)
+		{
+			copySize = dataLength - dataOffset;
+		}
+
+		pageHeader = (PageHeader) page;
+		pageHeader->pd_lower = SizeOfPageHeaderData + copySize;
+		memcpy(pageData, wholeFooter->data + dataOffset, copySize);
+
+		MarkBufferDirty(buffer);
+		actualBlockNumber = BufferGetBlockNumber(buffer);
+		if (loggingEnabled)
+		{
+			LOG_NEWPAGE_BUFFER(buffer);
+		}
+
+		END_CRIT_SECTION();
+
+		UnlockReleaseBuffer(buffer);
+
+		dataOffset += copySize;
+
+	}
+
+	/* all table footer is written, update the header page */
+	tableFooterMetadata = SerializeTableFooterMetadata(newStartingBlock, actualBlockCount);
+
+	START_CRIT_SECTION();
+	PageInit(headerPage, BLCKSZ, 0);
+	memcpy(headerPageContent, tableFooterMetadata->data, tableFooterMetadata->len);
+	headerBufferPageHeader->pd_lower = SizeOfPageHeaderData + tableFooterMetadata->len;
+	MarkBufferDirty(headerBuffer);
+
+	/*
+	 * Changes to headerBuffer is logged regardless of logging setting to enable
+	 * recovery after crash.
+	 */
+	LOG_NEWPAGE_BUFFER(headerBuffer);
+
+	END_CRIT_SECTION();
+
+	UnlockReleaseBuffer(headerBuffer);
 
 	pfree(tableFooterBuffer->data);
 	pfree(tableFooterBuffer);
 	pfree(postscriptBuffer->data);
 	pfree(postscriptBuffer);
+
+	pfree(wholeFooter->data);
+	pfree(wholeFooter);
+
+	pfree(tableFooterMetadata->data);
+	pfree(tableFooterMetadata);
 }
 
```

Comment:
caller had a memory context available. Removed all these pfree's and let the caller handle memory context.